brave/src/main/java/brave/Tracing.java

```diff
@@ -0,0 +1,215 @@
+     *                         .build();
+     *
+     * tracerBuilder.reporter(reporter);
+     * }</pre>
+     *
+     * <p>See https://github.com/openzipkin/zipkin-reporter-java
+     */
+    public Builder reporter(Reporter<zipkin.Span> reporter) {
+      if (reporter == null) throw new NullPointerException("reporter == null");
+      this.reporter = reporter;
+      return this;
+    }
+
+    /** See {@link Tracing#clock()} */
+    public Builder clock(Clock clock) {
+      if (clock == null) throw new NullPointerException("clock == null");
+      this.clock = clock;
+      return this;
+    }
+
+    /**
+     * Sampler is responsible for deciding if a particular trace should be "sampled", i.e. whether
+     * the overhead of tracing will occur and/or if a trace will be reported to Zipkin.
+     */
+    public Builder sampler(Sampler sampler) {
+      this.sampler = sampler;
+      return this;
+    }
+
+    /**
+     * Responsible for implementing {@link Tracer#currentSpan()} and {@link
+     * Tracer#withSpanInScope(Span)}. By default a simple thread-local is used. Override to support
+     * other mechanisms or to synchronize with other mechanisms such as SLF4J's MDC.
+     */
+    public Builder currentTraceContext(CurrentTraceContext currentTraceContext) {
+      this.currentTraceContext = currentTraceContext;
+      return this;
+    }
+
+    /** When true, new root spans will have 128-bit trace IDs. Defaults to false (64-bit) */
+    public Builder traceId128Bit(boolean traceId128Bit) {
+      this.traceId128Bit = traceId128Bit;
+      return this;
+    }
+
+    public Tracing build() {
+      if (clock == null) clock = Platform.get();
+      if (localEndpoint == null) {
+        localEndpoint = Platform.get().localEndpoint();
+        if (localServiceName != null) {
+          localEndpoint = localEndpoint.toBuilder().serviceName(localServiceName).build();
+        }
+      }
+      if (reporter == null) reporter = Platform.get();
+
+      return new Default(this);
+    }
+
+    Builder() {
+    }
+  }
+
+  static {
+    Internal.instance = new Internal() {
+      @Override public Long timestamp(Tracer tracer, TraceContext context) {
+        return tracer.recorder.timestamp(context);
+      }
+    };
+  }
+
+  static final class Default extends Tracing {
+    final Tracer tracer;
+    final Propagation.Factory propagationFactory;
+    final Propagation<String> stringPropagation;
+    final Clock clock;
+
+    Default(Builder builder) {
+      this.tracer = new Tracer(builder);
+      this.propagationFactory = builder.propagationFactory;
+      this.stringPropagation = builder.propagationFactory.create(Propagation.KeyFactory.STRING);
+      this.clock = builder.clock;
+      maybeSetCurrent();
+    }
+
+    @Override public Tracer tracer() {
+      return tracer;
+    }
+
+    @Override public Propagation.Factory propagationFactory() {
+      return propagationFactory;
+    }
```

Comment:
this is the only really new thing. everything else was pushing things from one place to another. should have called that out actually :)